jetty-http/src/main/java/org/eclipse/jetty/http/HttpCompliance.java

```diff
@@ -18,179 +18,394 @@
 
 package org.eclipse.jetty.http;
 
-import java.util.EnumSet;
-import java.util.HashMap;
-import java.util.Map;
-
-import org.eclipse.jetty.util.log.Log;
-import org.eclipse.jetty.util.log.Logger;
-
 /**
- * HTTP compliance modes for Jetty HTTP parsing and handling.
- * A Compliance mode consists of a set of {@link HttpComplianceSection}s which are applied
- * when the mode is enabled.
+ * HTTP (RFC) Spec Compliance behavior.
+ *
  * <p>
- * Currently the set of modes is an enum and cannot be dynamically extended, but future major releases may convert this
- * to a class. To modify modes there are four custom modes that can be modified by setting the property
- * <code>org.eclipse.jetty.http.HttpCompliance.CUSTOMn</code> (where 'n' is '0', '1', '2' or '3'), to a comma separated
- * list of sections.  The list should start with one of the following strings:<dl>
- * <dt>0</dt><dd>No {@link HttpComplianceSection}s</dd>
- * <dt>*</dt><dd>All {@link HttpComplianceSection}s</dd>
- * <dt>RFC2616</dt><dd>The set of {@link HttpComplianceSection}s application to https://tools.ietf.org/html/rfc2616,
- * but not https://tools.ietf.org/html/rfc7230</dd>
- * <dt>RFC7230</dt><dd>The set of {@link HttpComplianceSection}s application to https://tools.ietf.org/html/rfc7230</dd>
- * </dl>
- * The remainder of the list can contain then names of {@link HttpComplianceSection}s to include them in the mode, or prefixed
- * with a '-' to exclude thm from the mode.    Note that Jetty's modes may have some historic minor differences from the strict
- * RFC compliance, for example the <code>RFC2616_LEGACY</code> HttpCompliance is defined as 
- * <code>RFC2616,-FIELD_COLON,-METHOD_CASE_SENSITIVE</code>.
+ * Configurable behavior of HTTP/1.x with regards to compliance to current
+ * (<a href="https://tools.ietf.org/html/rfc7230">RFC 7230: HTTP/1.1</a>)
+ * and obsolete (<a href="https://tools.ietf.org/html/rfc2616">RFC 2616: HTTP/1.1</a>) Specs.
+ * </p>
+ *
  * <p>
- * Note also that the {@link EnumSet} return by {@link HttpCompliance#sections()} is mutable, so that modes may 
- * be altered in code and will affect all usages of the mode.
+ * To define a custom Compliance behavior, start with a {@link Builder} based on either
+ * {@link #rfc7230Builder()} or {@link #rfc2616Builder()}, modify it (or not), and then {@link Builder#build()} it.
+ * </p>
+ *
+ * <code>
+ * HttpCompliance compliance = HttpCompliance.rfc7230Builder()
+ *     .allowCaseInsensitiveFieldNames(true)
+ *     .requireColonAfterFieldName(true)
+ *     .allowHttp09(false)
+ *     .allowMultipleContentLengths(true)
+ *     .build();
+ * </code>
  */
-public enum HttpCompliance // TODO in Jetty-10 convert this enum to a class so that extra custom modes can be defined dynamically
+public final class HttpCompliance
 {
-    /** A Legacy compliance mode to match jetty's behavior prior to RFC2616 and RFC7230. 
-     */
-    LEGACY(sectionsBySpec("0,METHOD_CASE_SENSITIVE")), 
```

Comment:
I am +0 that we should maintain LEGACY... yes it is kind of a silly mode and badly named, but it represents the grab bag of various compliance modes that were notionally implemented by jetty prior to us have compliance modes.    There is still some limited value for that... at the very least in just documenting what the settings were.